ocaml/parsing/jane_syntax_parsing.ml

```diff
@@ -610,35 +715,67 @@ module Structure_item = Make_AST(struct
 end)
 
 (******************************************************************************)
-(** Generically lift our custom ASTs for our novel syntax from OCaml ASTs. *)
-
-module type Of_ast_parameters = sig
-  module AST : AST
-  type t
-  val of_ast_internal : Feature.t -> AST.ast -> t option
-end
 
-module Make_of_ast (Params : Of_ast_parameters) : sig
-  val of_ast : Params.AST.ast -> Params.t option
-end = struct
-  let of_ast ast =
-    let loc = Params.AST.location ast in
-    let raise_error err = raise (Error (loc, err)) in
-    match Params.AST.match_jane_syntax ast with
-    | Some ([name], ast) -> begin
-        match Feature.of_component name with
-        | Ok feat -> begin
-            match Params.of_ast_internal feat ast with
-            | Some ext_ast -> Some ext_ast
-            | None ->
-                raise_error (Wrong_syntactic_category(feat, Params.AST.plural))
+module AST = struct
+  type (_, _) t =
+    | Expression :
+        (Parsetree.expression, Parsetree.expression_desc With_attributes.t) t
+    | Pattern : (Parsetree.pattern, Parsetree.pattern_desc) t
+    | Module_type : (Parsetree.module_type, Parsetree.module_type_desc) t
+    | Signature_item :
+        (Parsetree.signature_item, Parsetree.signature_item_desc) t
+    | Structure_item :
+        (Parsetree.structure_item, Parsetree.structure_item_desc) t
+    | Core_type : (Parsetree.core_type, Parsetree.core_type_desc) t
+    | Constructor_argument : (Parsetree.core_type, Parsetree.core_type_desc) t
```

Comment:
Yes, it's to get type-directed disambiguation at use sites.

Reply:
I'm not sure this winds up being a win -- if you look at the use sites, it's now `___.AST.___ Core_type` rather than `___.Core_type.___`, isn't it?

Reply:
Oh, I meant something slightly different. I was answering a question "can we make `AST.t` be backed by a first-class module instead of a GADT", and of course the answer to that is yes (but that's not getting at what you're asking).

I think your question amounts to: why expose a type `AST.t` rather than modules of module type `AST`. That's mostly because I'd like to be more selective about exposing operations on `AST.t`. So, I'm going to do the `____.AST.___` refactor regardless -- it's just a matter of what the first argument to these functions are. I think a GADT is a nice choice because of disambiguation.

Reply:
Ah, now I see!  Yes, that makes sense to me.  Seems like a great choice all round.